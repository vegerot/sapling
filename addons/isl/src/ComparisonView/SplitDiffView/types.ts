/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {Comparison} from 'shared/Comparison';

/**
 * Context used to render SplitDiffView
 */
export type Context = {
  id: {path: string; comparison: Comparison};
  copy?: (s: string) => void;
  openFile?: () => unknown;
  openFileToLine?: (line: OneIndexedLineNumber) => unknown;
  collapsed: boolean;
  setCollapsed: (collapsed: boolean) => void;
  supportsExpandingContext: boolean;
  /**
   * Whether to render as a side-by-side diff view, or a unified view where deleted and added lines are interleaved.
   * TODO: make this controllable / configurable / responsive based on screen width
   */
  display: 'split' | 'unified';
};

export type OneIndexedLineNumber = Exclude<number, 0>;
