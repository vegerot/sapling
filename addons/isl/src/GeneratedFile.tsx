/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {RepoRelativePath} from './types';

import serverAPI from './ClientToServerAPI';
import {t} from './i18n';
import {writeAtom} from './jotaiUtils';
import {GeneratedStatus} from './types';
import {registerDisposable} from './utils';
import {atom, useAtomValue} from 'jotai';
import {useMemo} from 'react';
import {LRU} from 'shared/LRU';

export const genereatedFileCache = new LRU<RepoRelativePath, GeneratedStatus>(1500);

/** To avoid sending multiple redundant fetch requests, also save which files are being fetched right now */
const currentlyFetching = new Set<RepoRelativePath>();

/**
 * Generated files are cached in `generatedFileCache` LRU.
 * For historical reasons, the files are not an atom.
 * In order to allow rerender dependencies when we update file statuses,
 * store a generation index in recoil.
 * This state should generally be used through useGeneratedFileStatus helpers.
 */
const generatedFileGeneration = atom<number>(0);

registerDisposable(
  currentlyFetching,
  serverAPI.onMessageOfType('fetchedGeneratedStatuses', event => {
    for (const [path, status] of Object.entries(event.results)) {
      genereatedFileCache.set(path, status);
      currentlyFetching.delete(path);
    }
    writeAtom(generatedFileGeneration, old => old + 1);
  }),
  import.meta.hot,
);

export function useGeneratedFileStatus(path: RepoRelativePath): GeneratedStatus {
  useAtomValue(generatedFileGeneration); // update if we get new statuses
  const found = genereatedFileCache.get(path);
  if (found == null) {
    fetchMissingGeneratedFileStatuses([path]);
    return GeneratedStatus.Manual;
  }
  return found;
}

export function getGeneratedFilesFrom(paths: Array<RepoRelativePath>) {
  return Object.fromEntries(
    paths.map(path => [path, genereatedFileCache.get(path) ?? GeneratedStatus.Manual]),
  );
}

export function useGeneratedFileStatuses(
  paths: Array<RepoRelativePath>,
): Record<RepoRelativePath, GeneratedStatus> {
  const generation = useAtomValue(generatedFileGeneration); // update if we get new statuses

  fetchMissingGeneratedFileStatuses(paths);

  return useMemo(() => {
    return getGeneratedFilesFrom(paths);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [generation, paths]);
}

export function getCachedGeneratedFileStatuses(
  paths: Array<RepoRelativePath>,
): Record<RepoRelativePath, GeneratedStatus | undefined> {
  return Object.fromEntries(paths.map(path => [path, genereatedFileCache.get(path)]));
}

/**
 * Hint that this set of files are being used, any files missing from the generated files cache
 * should be checked on the server.
 * No-op if all files already in the cache.
 */
export function fetchMissingGeneratedFileStatuses(files: Array<RepoRelativePath>) {
  const notCached = files.filter(
    file => genereatedFileCache.get(file) == null && !currentlyFetching.has(file),
  );
  if (notCached.length > 0) {
    for (const file of notCached) {
      currentlyFetching.add(file);
    }
    serverAPI.postMessage({
      type: 'fetchGeneratedStatuses',
      paths: notCached,
    });
  }
}

export function generatedStatusToLabel(status: GeneratedStatus | undefined): string {
  if (status === GeneratedStatus.Generated) {
    return 'generated';
  } else if (status === GeneratedStatus.PartiallyGenerated) {
    return 'partial';
  } else {
    return 'manual';
  }
}

export function generatedStatusDescription(
  status: GeneratedStatus | undefined,
): string | undefined {
  if (status === GeneratedStatus.Generated) {
    return t('This file is generated');
  } else if (status === GeneratedStatus.PartiallyGenerated) {
    return t('This file is partially generated');
  } else {
    return undefined;
  }
}
