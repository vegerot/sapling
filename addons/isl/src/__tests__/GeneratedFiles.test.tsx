/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import App from '../App';
import {genereatedFileCache} from '../GeneratedFile';
import {__TEST__} from '../UncommittedChanges';
import {readAtom, writeAtom} from '../jotaiUtils';
import foundPlatform from '../platform';
import {ignoreRTL} from '../testQueries';
import {
  expectMessageSentToServer,
  simulateCommits,
  COMMIT,
  simulateUncommittedChangedFiles,
  closeCommitInfoSidebar,
  simulateRepoConnected,
  resetTestMessages,
  simulateMessageFromServer,
  openCommitInfoSidebar,
} from '../testUtils';
import {GeneratedStatus} from '../types';
import {fireEvent, render, screen, waitFor, act} from '@testing-library/react';

/** Generated `num` files, in the repeating pattern: genereated, partially generated, manual */
async function simulateGeneratedFiles(num: number) {
  const files = new Array(num).fill(null).map((_, i) => `file_${zeroPad(i)}.txt`);
  act(() => {
    simulateUncommittedChangedFiles({
      value: files.map(path => ({
        path,
        status: 'M',
      })),
    });
  });
  await waitFor(() => {
    expectMessageSentToServer({
      type: 'fetchGeneratedStatuses',
      paths: expect.anything(),
    });
  });
  act(() => {
    simulateMessageFromServer({
      type: 'fetchedGeneratedStatuses',
      results: Object.fromEntries(
        files.map((path, i) => [
          path,
          i % 3 === 0
            ? GeneratedStatus.Generated
            : i % 3 === 1
            ? GeneratedStatus.PartiallyGenerated
            : GeneratedStatus.Manual,
        ]),
      ),
    });
  });
}

function zeroPad(n: number): string {
  return ('000' + n.toString()).slice(-3);
}

describe('Generated Files', () => {
  beforeEach(() => {
    resetTestMessages();
    render(<App />);
    genereatedFileCache.clear();
    act(() => {
      simulateRepoConnected();
      closeCommitInfoSidebar();
      expectMessageSentToServer({
        type: 'subscribe',
        kind: 'smartlogCommits',
        subscriptionID: expect.anything(),
      });
      simulateCommits({
        value: [
          COMMIT('1', 'some public base', '0', {phase: 'public'}),
          COMMIT('a', 'My Commit', '1'),
          COMMIT('b', 'Another Commit', 'a', {isDot: true}),
        ],
      });
      expectMessageSentToServer({
        type: 'subscribe',
        kind: 'uncommittedChanges',
        subscriptionID: expect.anything(),
      });
    });
  });

  it('fetches generated files for uncommitted changes', async () => {
    await simulateGeneratedFiles(5);
    expectMessageSentToServer({
      type: 'fetchGeneratedStatuses',
      paths: ['file_000.txt', 'file_001.txt', 'file_002.txt', 'file_003.txt', 'file_004.txt'],
    });
  });

  it('Shows generated files in their own sections', async () => {
    await simulateGeneratedFiles(10);

    expect(screen.getByText(ignoreRTL('file_002.txt'))).toBeInTheDocument();
    expect(screen.getByText(ignoreRTL('file_005.txt'))).toBeInTheDocument();
    expect(screen.getByText(ignoreRTL('file_008.txt'))).toBeInTheDocument();
    expect(screen.getByText('Generated Files')).toBeInTheDocument();
    expect(screen.getByText('Partially Generated Files')).toBeInTheDocument();
  });

  function goToNextPage() {
    fireEvent.click(screen.getByTestId('changed-files-next-page'));
  }

  function expectHasPartiallyGeneratedFiles() {
    expect(screen.queryByText('Partially Generated Files')).toBeInTheDocument();
  }
  function expectHasGeneratedFiles() {
    expect(screen.queryByText('Generated Files')).toBeInTheDocument();
  }
  function expectNOTHasGeneratedFiles() {
    expect(screen.queryByText('Generated Files')).not.toBeInTheDocument();
  }

  function getChangedFiles() {
    const found = [...document.querySelectorAll('.changed-file-path-text')].map(e =>
      (e as HTMLElement).innerHTML.replace(/\u200E/g, ''),
    );
    return found;
  }

  it('Paginates generated files', async () => {
    await simulateGeneratedFiles(1200);
    // 1200 files, but 1000 files per fetched batch of generated statuses.
    // Sorted by status, that puts 1000/3 manual files, then 1000/3 partially generated, then 1000/3 generated,
    // then the remaining 200/3 manual, 200/3 partially generated, and 200/3 generated,
    // all in pages of 500 at a time.

    // first page is manual and partial
    expectHasPartiallyGeneratedFiles();
    expectNOTHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();

    // next page has partial and generated
    goToNextPage();
    expectHasPartiallyGeneratedFiles();
    expectHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();

    // next page has remaining files from all 3 types
    goToNextPage();
    expectHasPartiallyGeneratedFiles();
    expectHasGeneratedFiles();
    expect(getChangedFiles()).toMatchSnapshot();
  });

  it('Warns about too many files to fetch all generated statuses', async () => {
    await simulateGeneratedFiles(1001);
    expect(
      screen.getByText('There are more than 1000 files, some files may appear out of order'),
    ).toBeInTheDocument();
  });

  it('remembers expanded state', async () => {
    writeAtom(__TEST__.generatedFilesInitiallyExpanded, true);

    await simulateGeneratedFiles(1);

    expect(screen.getByText(ignoreRTL('file_000.txt'))).toBeInTheDocument();
    expect(screen.getByText('Generated Files')).toBeInTheDocument();
  });

  it('writes expanded state', async () => {
    expect(readAtom(__TEST__.generatedFilesInitiallyExpanded)).toEqual(false);

    await simulateGeneratedFiles(1);

    fireEvent.click(screen.getByText('Generated Files'));

    expect(readAtom(__TEST__.generatedFilesInitiallyExpanded)).toEqual(true);
  });

  it('clears generated files cache on refresh click', async () => {
    act(() => {
      simulateUncommittedChangedFiles({
        value: [
          {
            path: 'file.txt',
            status: 'M',
          },
        ],
      });
    });
    await waitFor(() => {
      expectMessageSentToServer({
        type: 'fetchGeneratedStatuses',
        paths: ['file.txt'],
      });
    });
    act(() => {
      simulateMessageFromServer({
        type: 'fetchedGeneratedStatuses',
        results: Object.fromEntries([['file.txt', GeneratedStatus.Manual]]),
      });
    });

    expect(screen.queryByText('Generated Files')).not.toBeInTheDocument();

    act(() => {
      fireEvent.click(screen.getByTestId('refresh-button'));
    });
    await waitFor(() => {
      expectMessageSentToServer({
        type: 'fetchGeneratedStatuses',
        paths: ['file.txt'],
      });
    });
    act(() => {
      simulateMessageFromServer({
        type: 'fetchedGeneratedStatuses',
        results: Object.fromEntries([['file.txt', GeneratedStatus.Generated]]),
      });
    });

    expect(screen.getByText('Generated Files')).toBeInTheDocument();
  });

  describe('Open All Files', () => {
    beforeEach(() => act(() => openCommitInfoSidebar()));
    async function simulateCommitWithFiles(files: Record<string, GeneratedStatus>) {
      act(() => {
        simulateCommits({
          value: [
            COMMIT('1', 'some public base', '0', {phase: 'public'}),
            COMMIT('a', 'Commit A', '1', {
              isDot: true,
              totalFileCount: 3,
              filesSample: Object.keys(files).map(file => ({path: file, status: 'M'})),
            }),
          ],
        });
      });
      await waitFor(() => {
        expectMessageSentToServer({
          type: 'fetchGeneratedStatuses',
          paths: expect.anything(),
        });
      });
      act(() => {
        simulateMessageFromServer({
          type: 'fetchedGeneratedStatuses',
          results: files,
        });
      });
    }

    it('No generated files, opens all files', async () => {
      const openSpy = jest.spyOn(foundPlatform, 'openFiles').mockImplementation(() => {});
      await simulateCommitWithFiles({
        'file_partial.txt': GeneratedStatus.PartiallyGenerated,
        'file_manual.txt': GeneratedStatus.Manual,
      });

      fireEvent.click(screen.getByText('Open All Files'));
      expect(openSpy).toHaveBeenCalledTimes(1);
      expect(openSpy).toHaveBeenCalledWith(['file_partial.txt', 'file_manual.txt']);
    });

    it('Some generated files, opens all non-generated files', async () => {
      const openSpy = jest.spyOn(foundPlatform, 'openFiles').mockImplementation(() => {});
      await simulateCommitWithFiles({
        'file_gen.txt': GeneratedStatus.Generated,
        'file_partial.txt': GeneratedStatus.PartiallyGenerated,
        'file_manual.txt': GeneratedStatus.Manual,
      });

      fireEvent.click(screen.getByText('Open Non-Generated Files'));
      expect(openSpy).toHaveBeenCalledTimes(1);
      expect(openSpy).toHaveBeenCalledWith(['file_partial.txt', 'file_manual.txt']);
    });

    it('All generated files, opens all files', async () => {
      const openSpy = jest.spyOn(foundPlatform, 'openFiles').mockImplementation(() => {});
      await simulateCommitWithFiles({
        'file_gen1.txt': GeneratedStatus.Generated,
        'file_gen2.txt': GeneratedStatus.Generated,
      });

      fireEvent.click(screen.getByText('Open All Files'));
      expect(openSpy).toHaveBeenCalledTimes(1);
      expect(openSpy).toHaveBeenCalledWith(['file_gen1.txt', 'file_gen2.txt']);
    });
  });
});
