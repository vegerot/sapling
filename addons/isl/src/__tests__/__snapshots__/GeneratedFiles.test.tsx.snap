// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generated Files Paginates generated files 1`] = `
[
  "file_001.txt",
  "file_004.txt",
  "file_007.txt",
  "file_010.txt",
  "file_013.txt",
  "file_016.txt",
  "file_019.txt",
  "file_022.txt",
  "file_025.txt",
  "file_028.txt",
  "file_031.txt",
  "file_034.txt",
  "file_037.txt",
  "file_040.txt",
  "file_043.txt",
  "file_046.txt",
  "file_049.txt",
  "file_052.txt",
  "file_055.txt",
  "file_058.txt",
  "file_061.txt",
  "file_064.txt",
  "file_067.txt",
  "file_070.txt",
  "file_073.txt",
  "file_076.txt",
  "file_079.txt",
  "file_082.txt",
  "file_085.txt",
  "file_088.txt",
  "file_091.txt",
  "file_094.txt",
  "file_097.txt",
  "file_100.txt",
  "file_103.txt",
  "file_106.txt",
  "file_109.txt",
  "file_112.txt",
  "file_115.txt",
  "file_118.txt",
  "file_121.txt",
  "file_124.txt",
  "file_127.txt",
  "file_130.txt",
  "file_133.txt",
  "file_136.txt",
  "file_139.txt",
  "file_142.txt",
  "file_145.txt",
  "file_148.txt",
  "file_151.txt",
  "file_154.txt",
  "file_157.txt",
  "file_160.txt",
  "file_163.txt",
  "file_166.txt",
  "file_169.txt",
  "file_172.txt",
  "file_175.txt",
  "file_178.txt",
  "file_181.txt",
  "file_184.txt",
  "file_187.txt",
  "file_190.txt",
  "file_193.txt",
  "file_196.txt",
  "file_199.txt",
  "file_200.txt",
  "file_203.txt",
  "file_206.txt",
  "file_209.txt",
  "file_212.txt",
  "file_215.txt",
  "file_218.txt",
  "file_221.txt",
  "file_224.txt",
  "file_227.txt",
  "file_230.txt",
  "file_233.txt",
  "file_236.txt",
  "file_239.txt",
  "file_242.txt",
  "file_245.txt",
  "file_248.txt",
  "file_251.txt",
  "file_254.txt",
  "file_257.txt",
  "file_260.txt",
  "file_263.txt",
  "file_266.txt",
  "file_269.txt",
  "file_272.txt",
  "file_275.txt",
  "file_278.txt",
  "file_281.txt",
  "file_284.txt",
  "file_287.txt",
  "file_290.txt",
  "file_293.txt",
  "file_296.txt",
  "file_299.txt",
  "file_302.txt",
  "file_305.txt",
  "file_308.txt",
  "file_311.txt",
  "file_314.txt",
  "file_317.txt",
  "file_320.txt",
  "file_323.txt",
  "file_326.txt",
  "file_329.txt",
  "file_332.txt",
  "file_335.txt",
  "file_338.txt",
  "file_341.txt",
  "file_344.txt",
  "file_347.txt",
  "file_350.txt",
  "file_353.txt",
  "file_356.txt",
  "file_359.txt",
  "file_362.txt",
  "file_365.txt",
  "file_368.txt",
  "file_371.txt",
  "file_374.txt",
  "file_377.txt",
  "file_380.txt",
  "file_383.txt",
  "file_386.txt",
  "file_389.txt",
  "file_392.txt",
  "file_395.txt",
  "file_398.txt",
  "file_401.txt",
  "file_404.txt",
  "file_407.txt",
  "file_410.txt",
  "file_413.txt",
  "file_416.txt",
  "file_419.txt",
  "file_422.txt",
  "file_425.txt",
  "file_428.txt",
  "file_431.txt",
  "file_434.txt",
  "file_437.txt",
  "file_440.txt",
  "file_443.txt",
  "file_446.txt",
  "file_449.txt",
  "file_452.txt",
  "file_455.txt",
  "file_458.txt",
  "file_461.txt",
  "file_464.txt",
  "file_467.txt",
  "file_470.txt",
  "file_473.txt",
  "file_476.txt",
  "file_479.txt",
  "file_482.txt",
  "file_485.txt",
  "file_488.txt",
  "file_491.txt",
  "file_494.txt",
  "file_497.txt",
  "file_500.txt",
  "file_503.txt",
  "file_506.txt",
  "file_509.txt",
  "file_512.txt",
  "file_515.txt",
  "file_518.txt",
  "file_521.txt",
  "file_524.txt",
  "file_527.txt",
  "file_530.txt",
  "file_533.txt",
  "file_536.txt",
  "file_539.txt",
  "file_542.txt",
  "file_545.txt",
  "file_548.txt",
  "file_551.txt",
  "file_554.txt",
  "file_557.txt",
  "file_560.txt",
  "file_563.txt",
  "file_566.txt",
  "file_569.txt",
  "file_572.txt",
  "file_575.txt",
  "file_578.txt",
  "file_581.txt",
  "file_584.txt",
  "file_587.txt",
  "file_590.txt",
  "file_593.txt",
  "file_596.txt",
  "file_599.txt",
  "file_602.txt",
  "file_605.txt",
  "file_608.txt",
  "file_611.txt",
  "file_614.txt",
  "file_617.txt",
  "file_620.txt",
  "file_623.txt",
  "file_626.txt",
  "file_629.txt",
  "file_632.txt",
  "file_635.txt",
  "file_638.txt",
  "file_641.txt",
  "file_644.txt",
  "file_647.txt",
  "file_650.txt",
  "file_653.txt",
  "file_656.txt",
  "file_659.txt",
  "file_662.txt",
  "file_665.txt",
  "file_668.txt",
  "file_671.txt",
  "file_674.txt",
  "file_677.txt",
  "file_680.txt",
  "file_683.txt",
  "file_686.txt",
  "file_689.txt",
  "file_692.txt",
  "file_695.txt",
  "file_698.txt",
  "file_701.txt",
  "file_704.txt",
  "file_707.txt",
  "file_710.txt",
  "file_713.txt",
  "file_716.txt",
  "file_719.txt",
  "file_722.txt",
  "file_725.txt",
  "file_728.txt",
  "file_731.txt",
  "file_734.txt",
  "file_737.txt",
  "file_740.txt",
  "file_743.txt",
  "file_746.txt",
  "file_749.txt",
  "file_752.txt",
  "file_755.txt",
  "file_758.txt",
  "file_761.txt",
  "file_764.txt",
  "file_767.txt",
  "file_770.txt",
  "file_773.txt",
  "file_776.txt",
  "file_779.txt",
  "file_782.txt",
  "file_785.txt",
  "file_788.txt",
  "file_791.txt",
  "file_794.txt",
  "file_797.txt",
  "file_800.txt",
  "file_803.txt",
  "file_806.txt",
  "file_809.txt",
  "file_812.txt",
  "file_815.txt",
  "file_818.txt",
  "file_821.txt",
  "file_824.txt",
  "file_827.txt",
  "file_830.txt",
  "file_833.txt",
  "file_836.txt",
  "file_839.txt",
  "file_842.txt",
  "file_845.txt",
  "file_848.txt",
  "file_851.txt",
  "file_854.txt",
  "file_857.txt",
  "file_860.txt",
  "file_863.txt",
  "file_866.txt",
  "file_869.txt",
  "file_872.txt",
  "file_875.txt",
  "file_878.txt",
  "file_881.txt",
  "file_884.txt",
  "file_887.txt",
  "file_890.txt",
  "file_893.txt",
  "file_896.txt",
  "file_899.txt",
  "file_902.txt",
  "file_905.txt",
  "file_908.txt",
  "file_911.txt",
  "file_914.txt",
  "file_917.txt",
  "file_920.txt",
  "file_923.txt",
  "file_926.txt",
  "file_929.txt",
  "file_932.txt",
  "file_935.txt",
  "file_938.txt",
  "file_941.txt",
  "file_944.txt",
  "file_947.txt",
  "file_950.txt",
  "file_953.txt",
  "file_956.txt",
  "file_959.txt",
  "file_962.txt",
  "file_965.txt",
  "file_968.txt",
  "file_971.txt",
  "file_974.txt",
  "file_977.txt",
  "file_980.txt",
  "file_983.txt",
  "file_986.txt",
  "file_989.txt",
  "file_992.txt",
  "file_995.txt",
  "file_998.txt",
  "file_000.txt",
  "file_003.txt",
  "file_006.txt",
  "file_009.txt",
  "file_012.txt",
  "file_015.txt",
  "file_018.txt",
  "file_021.txt",
  "file_024.txt",
  "file_027.txt",
  "file_030.txt",
  "file_033.txt",
  "file_036.txt",
  "file_039.txt",
  "file_042.txt",
  "file_045.txt",
  "file_048.txt",
  "file_051.txt",
  "file_054.txt",
  "file_057.txt",
  "file_060.txt",
  "file_063.txt",
  "file_066.txt",
  "file_069.txt",
  "file_072.txt",
  "file_075.txt",
  "file_078.txt",
  "file_081.txt",
  "file_084.txt",
  "file_087.txt",
  "file_090.txt",
  "file_093.txt",
  "file_096.txt",
  "file_099.txt",
  "file_102.txt",
  "file_105.txt",
  "file_108.txt",
  "file_111.txt",
  "file_114.txt",
  "file_117.txt",
  "file_120.txt",
  "file_123.txt",
  "file_126.txt",
  "file_129.txt",
  "file_132.txt",
  "file_135.txt",
  "file_138.txt",
  "file_141.txt",
  "file_144.txt",
  "file_147.txt",
  "file_150.txt",
  "file_153.txt",
  "file_156.txt",
  "file_159.txt",
  "file_162.txt",
  "file_165.txt",
  "file_168.txt",
  "file_171.txt",
  "file_174.txt",
  "file_177.txt",
  "file_180.txt",
  "file_183.txt",
  "file_186.txt",
  "file_189.txt",
  "file_192.txt",
  "file_195.txt",
  "file_198.txt",
  "file_202.txt",
  "file_205.txt",
  "file_208.txt",
  "file_211.txt",
  "file_214.txt",
  "file_217.txt",
  "file_220.txt",
  "file_223.txt",
  "file_226.txt",
  "file_229.txt",
  "file_232.txt",
  "file_235.txt",
  "file_238.txt",
  "file_241.txt",
  "file_244.txt",
  "file_247.txt",
  "file_250.txt",
  "file_253.txt",
  "file_256.txt",
  "file_259.txt",
  "file_262.txt",
  "file_265.txt",
  "file_268.txt",
  "file_271.txt",
  "file_274.txt",
  "file_277.txt",
  "file_280.txt",
  "file_283.txt",
  "file_286.txt",
  "file_289.txt",
  "file_292.txt",
  "file_295.txt",
  "file_298.txt",
  "file_301.txt",
  "file_304.txt",
  "file_307.txt",
  "file_310.txt",
  "file_313.txt",
  "file_316.txt",
  "file_319.txt",
  "file_322.txt",
  "file_325.txt",
  "file_328.txt",
  "file_331.txt",
  "file_334.txt",
  "file_337.txt",
  "file_340.txt",
  "file_343.txt",
  "file_346.txt",
  "file_349.txt",
  "file_352.txt",
  "file_355.txt",
  "file_358.txt",
  "file_361.txt",
  "file_364.txt",
  "file_367.txt",
  "file_370.txt",
  "file_373.txt",
  "file_376.txt",
  "file_379.txt",
  "file_382.txt",
  "file_385.txt",
  "file_388.txt",
  "file_391.txt",
  "file_394.txt",
  "file_397.txt",
  "file_400.txt",
  "file_403.txt",
  "file_406.txt",
  "file_409.txt",
  "file_412.txt",
  "file_415.txt",
  "file_418.txt",
  "file_421.txt",
  "file_424.txt",
  "file_427.txt",
  "file_430.txt",
  "file_433.txt",
  "file_436.txt",
  "file_439.txt",
  "file_442.txt",
  "file_445.txt",
  "file_448.txt",
  "file_451.txt",
  "file_454.txt",
  "file_457.txt",
  "file_460.txt",
  "file_463.txt",
  "file_466.txt",
  "file_469.txt",
  "file_472.txt",
  "file_475.txt",
  "file_478.txt",
  "file_481.txt",
  "file_484.txt",
  "file_487.txt",
  "file_490.txt",
  "file_493.txt",
  "file_496.txt",
]
`;

exports[`Generated Files Paginates generated files 2`] = `
[
  "file_499.txt",
  "file_502.txt",
  "file_505.txt",
  "file_508.txt",
  "file_511.txt",
  "file_514.txt",
  "file_517.txt",
  "file_520.txt",
  "file_523.txt",
  "file_526.txt",
  "file_529.txt",
  "file_532.txt",
  "file_535.txt",
  "file_538.txt",
  "file_541.txt",
  "file_544.txt",
  "file_547.txt",
  "file_550.txt",
  "file_553.txt",
  "file_556.txt",
  "file_559.txt",
  "file_562.txt",
  "file_565.txt",
  "file_568.txt",
  "file_571.txt",
  "file_574.txt",
  "file_577.txt",
  "file_580.txt",
  "file_583.txt",
  "file_586.txt",
  "file_589.txt",
  "file_592.txt",
  "file_595.txt",
  "file_598.txt",
  "file_601.txt",
  "file_604.txt",
  "file_607.txt",
  "file_610.txt",
  "file_613.txt",
  "file_616.txt",
  "file_619.txt",
  "file_622.txt",
  "file_625.txt",
  "file_628.txt",
  "file_631.txt",
  "file_634.txt",
  "file_637.txt",
  "file_640.txt",
  "file_643.txt",
  "file_646.txt",
  "file_649.txt",
  "file_652.txt",
  "file_655.txt",
  "file_658.txt",
  "file_661.txt",
  "file_664.txt",
  "file_667.txt",
  "file_670.txt",
  "file_673.txt",
  "file_676.txt",
  "file_679.txt",
  "file_682.txt",
  "file_685.txt",
  "file_688.txt",
  "file_691.txt",
  "file_694.txt",
  "file_697.txt",
  "file_700.txt",
  "file_703.txt",
  "file_706.txt",
  "file_709.txt",
  "file_712.txt",
  "file_715.txt",
  "file_718.txt",
  "file_721.txt",
  "file_724.txt",
  "file_727.txt",
  "file_730.txt",
  "file_733.txt",
  "file_736.txt",
  "file_739.txt",
  "file_742.txt",
  "file_745.txt",
  "file_748.txt",
  "file_751.txt",
  "file_754.txt",
  "file_757.txt",
  "file_760.txt",
  "file_763.txt",
  "file_766.txt",
  "file_769.txt",
  "file_772.txt",
  "file_775.txt",
  "file_778.txt",
  "file_781.txt",
  "file_784.txt",
  "file_787.txt",
  "file_790.txt",
  "file_793.txt",
  "file_796.txt",
  "file_799.txt",
  "file_802.txt",
  "file_805.txt",
  "file_808.txt",
  "file_811.txt",
  "file_814.txt",
  "file_817.txt",
  "file_820.txt",
  "file_823.txt",
  "file_826.txt",
  "file_829.txt",
  "file_832.txt",
  "file_835.txt",
  "file_838.txt",
  "file_841.txt",
  "file_844.txt",
  "file_847.txt",
  "file_850.txt",
  "file_853.txt",
  "file_856.txt",
  "file_859.txt",
  "file_862.txt",
  "file_865.txt",
  "file_868.txt",
  "file_871.txt",
  "file_874.txt",
  "file_877.txt",
  "file_880.txt",
  "file_883.txt",
  "file_886.txt",
  "file_889.txt",
  "file_892.txt",
  "file_895.txt",
  "file_898.txt",
  "file_901.txt",
  "file_904.txt",
  "file_907.txt",
  "file_910.txt",
  "file_913.txt",
  "file_916.txt",
  "file_919.txt",
  "file_922.txt",
  "file_925.txt",
  "file_928.txt",
  "file_931.txt",
  "file_934.txt",
  "file_937.txt",
  "file_940.txt",
  "file_943.txt",
  "file_946.txt",
  "file_949.txt",
  "file_952.txt",
  "file_955.txt",
  "file_958.txt",
  "file_961.txt",
  "file_964.txt",
  "file_967.txt",
  "file_970.txt",
  "file_973.txt",
  "file_976.txt",
  "file_979.txt",
  "file_982.txt",
  "file_985.txt",
  "file_988.txt",
  "file_991.txt",
  "file_994.txt",
  "file_997.txt",
]
`;

exports[`Generated Files Paginates generated files 3`] = `
[
  "file_001.txt",
  "file_004.txt",
  "file_007.txt",
  "file_010.txt",
  "file_013.txt",
  "file_016.txt",
  "file_019.txt",
  "file_022.txt",
  "file_025.txt",
  "file_028.txt",
  "file_031.txt",
  "file_034.txt",
  "file_037.txt",
  "file_040.txt",
  "file_043.txt",
  "file_046.txt",
  "file_049.txt",
  "file_052.txt",
  "file_055.txt",
  "file_058.txt",
  "file_061.txt",
  "file_064.txt",
  "file_067.txt",
  "file_070.txt",
  "file_073.txt",
  "file_076.txt",
  "file_079.txt",
  "file_082.txt",
  "file_085.txt",
  "file_088.txt",
  "file_091.txt",
  "file_094.txt",
  "file_097.txt",
  "file_100.txt",
  "file_103.txt",
  "file_106.txt",
  "file_109.txt",
  "file_112.txt",
  "file_115.txt",
  "file_118.txt",
  "file_121.txt",
  "file_124.txt",
  "file_127.txt",
  "file_130.txt",
  "file_133.txt",
  "file_136.txt",
  "file_139.txt",
  "file_142.txt",
  "file_145.txt",
  "file_148.txt",
  "file_151.txt",
  "file_154.txt",
  "file_157.txt",
  "file_160.txt",
  "file_163.txt",
  "file_166.txt",
  "file_169.txt",
  "file_172.txt",
  "file_175.txt",
  "file_178.txt",
  "file_181.txt",
  "file_184.txt",
  "file_187.txt",
  "file_190.txt",
  "file_193.txt",
  "file_196.txt",
  "file_199.txt",
  "file_000.txt",
  "file_003.txt",
  "file_006.txt",
  "file_009.txt",
  "file_012.txt",
  "file_015.txt",
  "file_018.txt",
  "file_021.txt",
  "file_024.txt",
  "file_027.txt",
  "file_030.txt",
  "file_033.txt",
  "file_036.txt",
  "file_039.txt",
  "file_042.txt",
  "file_045.txt",
  "file_048.txt",
  "file_051.txt",
  "file_054.txt",
  "file_057.txt",
  "file_060.txt",
  "file_063.txt",
  "file_066.txt",
  "file_069.txt",
  "file_072.txt",
  "file_075.txt",
  "file_078.txt",
  "file_081.txt",
  "file_084.txt",
  "file_087.txt",
  "file_090.txt",
  "file_093.txt",
  "file_096.txt",
  "file_099.txt",
  "file_102.txt",
  "file_105.txt",
  "file_108.txt",
  "file_111.txt",
  "file_114.txt",
  "file_117.txt",
  "file_120.txt",
  "file_123.txt",
  "file_126.txt",
  "file_129.txt",
  "file_132.txt",
  "file_135.txt",
  "file_138.txt",
  "file_141.txt",
  "file_144.txt",
  "file_147.txt",
  "file_150.txt",
  "file_153.txt",
  "file_156.txt",
  "file_159.txt",
  "file_162.txt",
  "file_165.txt",
  "file_168.txt",
  "file_171.txt",
  "file_174.txt",
  "file_177.txt",
  "file_180.txt",
  "file_183.txt",
  "file_186.txt",
  "file_189.txt",
  "file_192.txt",
  "file_195.txt",
  "file_198.txt",
]
`;
